### PR TITLE
Call node destructors when freeing the Context.

### PR DESCRIPTION
**Pull request by nweston**
Adds a subclass of BumpPtrAllocatorImpl which only allocates a single
type, allowing it to walk the slabs and destroy objects during its own
destructor.

Context now has a separate allocator for each node type, with a
templated Allocate method which is specialized to use the correct
allocator.

The `Create` methods now call `Context::Allocate` explicitly, and use the built-in placement new rather than our own version. So this code:
```c++
return new (C) Module(C);
```
becomes this:
```c++
return new (C.Allocate<Module>()) Module(C);
```
This allows us to pass the type information down to the `Context`, so it can pick the right allocator.

There are two things which I don't really like about this design:

1. `DtorBumpPtrAllocator` is very dependent on the details of `BumpPtrAllocatorImpl`. But the public API is insufficient for our purposes, and it seems better to use an existing allocator which is presumably well-tested than to start from scratch. And in any event we aren't likely to make a lot of changes to this code. So I think this is ultimately the right way to go.

2. `DtorBumpPtrAllocator` is responsible for calling the destructors, but the various `Create` methods determine the type of the allocated object. So essentially, the two sides of the process are handled at different layers. One can imagine a copy-paste error leading to code like:
```c++
return new (C.Allocate<Module>()) MyNode(C);
```

This won't be detected at compile time and could lead to subtle errors. But this seems to be an unavoidable consequence of the way placement new works.

### CODE REVIEW

File: include/gtirb/Allocator.hpp [hunk 1]
```diff
@@ -264,6 +264,44 @@ class BumpPtrAllocatorImpl {
   }
 };
 
+/// Allocate memory for a specific type in an ever-growing pool, and call
+/// destructors for all allocated objects when freeing the pool.
+template <class T, size_t SlabSize = 4096, size_t SizeThreshold = SlabSize>
```
Comment (AaronBallman) [inline include/gtirb/Allocator.hpp]: Should we make the `SlabSize` a multiple of `sizeof(T)`?

File: include/gtirb/Allocator.hpp [hunk 2]
```diff
@@ -264,6 +264,44 @@ class BumpPtrAllocatorImpl {
...
+class DtorBumpPtrAllocator : BumpPtrAllocatorImpl<SlabSize, SizeThreshold> {
```
Comment (AaronBallman) [inline include/gtirb/Allocator.hpp]: The way that I'm used to seeing this done is more like `SpecificBumpPtrAllocator` (https://llvm.org/viewvc/llvm-project/llvm/trunk/include/llvm/Support/Allocator.h?view=markup#l415) where it composes rather than inherits. Did you consider such a design?

File: src/Context.cpp [hunk 3]
```diff
@@ -13,10 +13,23 @@
...
+#include <gtirb/IR.hpp>
+#include <gtirb/ImageByteMap.hpp>
+#include <gtirb/Module.hpp>
 #include <gtirb/Node.hpp>
+#include <gtirb/Section.hpp>
+#include <gtirb/Symbol.hpp>
 
 using namespace gtirb;
 
+// By moving these declarations here, we avoid instantiating the default
```
Comment (AaronBallman) [inline src/Context.cpp]: Did you test this with MSVC as well as Clang?

Comment (nweston) [reply src/Context.cpp]: Yes.

File: include/gtirb/Allocator.hpp (hunk 2 above)
Comment (nweston) [reply include/gtirb/Allocator.hpp]: Oh, I wish I'd know that existed. I should probably throw out my allocator and use that instead, since it goes along with the `BumpPtrAllocator` that we've already copied.

**Comment by nweston** [comment]
I've replaced my allocator with `SpecificBumpPtrAllocator` from LLVM, which resolves all the comments about the specifics of the allocator. And I reworked the `Create` methods as we discussed yesterday.

File: include/gtirb/Allocator.hpp [hunk 4]
```diff
@@ -201,7 +206,7 @@ class BumpPtrAllocatorImpl {
 
   void setRedZoneSize(size_t NewSize) { RedZoneSize = NewSize; }
 
-private:
```
Comment (AaronBallman) [inline include/gtirb/Allocator.hpp]: Is this change still needed?

File: include/gtirb/Context.hpp [hunk 5]
```diff
@@ -64,17 +81,21 @@ class GTIRB_EXPORT_API Context {
...
+  /// \tparam T   The type of object for which to allocate memory.
   ///
-  /// \return The newly allocated memory, suitably aligned for the given
-  /// alignment. Will return nullptr if the allocation cannot be honored.
-  void* Allocate(size_t Size, size_t Align) const {
-    return Allocator.Allocate(Size, Align);
-  }
+  /// \return The newly allocated memory, suitably sized for the given
+  /// type. Will return nullptr if the allocation cannot be honored.
+  template <class T> void* Allocate() const;
```
Comment (AaronBallman) [inline include/gtirb/Context.hpp]: Can this be made private?

File: include/gtirb/Symbol.hpp [hunk 6]
```diff
@@ -217,7 +217,9 @@ class GTIRB_EXPORT_API Symbol : public Node {
   /// \param C  The Context in which this object will be held.
   ///
   /// \return The newly created object.
-  static Symbol* Create(Context& C) { return new (C) Symbol(C); }
+  static Symbol* Create(Context& C) {
+    return new (C.Allocate<Symbol>()) Symbol(C);
```
Comment (AaronBallman) [inline include/gtirb/Symbol.hpp]: Shouldn't this be using `C.Create<Symbol>()`? (Same below.)

File: include/gtirb/Allocator.hpp (hunk 4 above)
Comment (nweston) [reply include/gtirb/Allocator.hpp]: No.

File: include/gtirb/Context.hpp (hunk 5 above)
Comment (nweston) [reply include/gtirb/Context.hpp]: Yes.

**Comment by nweston** [comment]
All fixed now.
